### DISCUSSION TITLE
TS: Simple defaultValues Record<MyFieldNames, string> not assignable to UnpackNestedValue<DeepPartial<Record<TFieldNames, string>>>

### DISCUSSION THREAD

**Discussion by devinrhode2** [react-hook-form/react-hook-form#3713] [Q&A]
I'd like to avoid doing a typecast when I'm passing a simple object to `useForm({ defaultValues })`.

Right now I have something like this:
![RHF default values crap2](https://user-images.githubusercontent.com/539816/102436630-a34db300-3fde-11eb-9dfe-9dc40eb28edf.png)
![RHF Default Values crap](https://user-images.githubusercontent.com/539816/102436616-9a5ce180-3fde-11eb-98c2-dbcef081987b.png)

As a work-around I am just type casting to `UnpackNestedValue<DeepPartial<typeof defaultValues>>`, but would rather not have to use this type casting.

`defaultValues` is a simple object like `{ firstName: string, lastName: string }`

**Comment by bluebill1049** [comment]
can you share a codesandbox?

**Reply by devinrhode2** [reply]
Yes, this one wasn't too hard! :) https://codesandbox.io/s/frosty-mccarthy-cdqn6?file=/src/App.tsx:0-718

**Reply by devinrhode2** [reply]
For permanence, it's just this:
```ts
import * as React from "react";
import { useForm } from "react-hook-form";

interface RhfFormProps<TFieldNames extends string> {
  staticConfig: {
    defaultValuesSchema: Record<TFieldNames, string>;
    preFilledValues: Record<TFieldNames, string>; // may actually be a partial, but that's beside the point.
  };
}

const RhfForm = <TFieldNames extends string>({
  // destructure our things
  staticConfig: { defaultValuesSchema, preFilledValues }
}: RhfFormProps<TFieldNames>) => {
  const defaultValues = {
    ...defaultValuesSchema,
    ...preFilledValues
  };
  const { register, errors } = useForm({
    defaultValues
  });
};
```

**Reply by devinrhode2** [reply]
Here's a simpler more focused example: https://codesandbox.io/s/defaultvalues-typing-unpacknestedvalue-yowu0?file=/src/App.tsx

```ts
/* eslint-disable @typescript-eslint/no-unused-vars, @typescript-eslint/no-use-before-define */
import * as React from "react";
import { useForm, DeepPartial, UnpackNestedValue } from "react-hook-form";

export default function App<TFieldNames extends string>(
  // This works (simple hard-coding):
  // _defaultValues: { "firstName": "SpongeBob" }

  // This does not work: (fails when passed to `useForm` below)
  _defaultValues: Record<TFieldNames, "SpongeBob">

  // Index type also fails:
  // _defaultValues: { [x in TFieldNames]: "SpongeBob" }
  // (Changing value to "" or `string` doesn't help either)
) {
  // Comment in one of these assignments: (in the codesandbox)
  const defaultValues = _defaultValues; // exposes error in useForm call
  // 1. Patch with simple type cast:
  // const defaultValues = _defaultValues as UnpackNestedValue<DeepPartial<typeof schema>>;
  // 2. Patch with utility:
  // const defaultValues = PatchRHFDefaultValuesType(_defaultValues);
  const { register, errors } = useForm({
    defaultValues
  });
  return "PROTIP: Shrink this render pane :)";
}

// Utility function to patch types:
const PatchRHFDefaultValuesType = <TDefaultValues extends unknown>(
  yourDefaultValues: TDefaultValues
  // If you have a flat form, but all fields are prefixed one level
  // with a formId like so: `MyFormId.firstName`, then use this arg type:
  // yourDefaultValues: Record<string, TDefaultValues>
): UnpackNestedValue<DeepPartial<TDefaultValues>> => {
  return yourDefaultValues as UnpackNestedValue<DeepPartial<TDefaultValues>>;
};
```

Overall the `UnpackNestedValue` stuff seems pretty buggy, but it's very advanced TS. I sense it has something to do with allowing namespaced form field names: `<input name="MyForm.firstName"`, but I'm not sure how to approach possibly improving this

**Reply by devinrhode2** [reply]
Small aside, but may be quite helpful, I noticed that in `src/fields/types.ts` we have `FieldValues` set like so:
```ts
export type FieldValues = Record<string, any>;
```
I've personally had a little bit of clashing with this type I think, but it's quite a rabbit hole to try and explain. Shouldn't FieldValues be more like `Record<string, boolean | number | string | null | undefined>` ? Or it'd be recursive like:
```ts
export type PrimitiveValue = boolean | number | string | null | undefined
export type PrimitiveFieldValues = Record<string, PrimitiveValue>;
export type FieldValues = Record<string, PrimitiveValue | PrimitiveFieldValues>
```
I've never written recursive types yet, but typescript doesn't seem to complain: https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBDAnmYcAKUCWBbTNMBuwAagIYA2ArqgLxwBGEE5wpAdnAD5xuXb3AoXOAGcYWNgHNhvcuWGU2AE2AAzTG2BKAUKEiwEyVBhx5CwAGKZg5JWSrARcOgCVgAY2hKAPGImSAGnQsXHwie2oAPgBuXXBoeCQUOCsbOwpqJ1cPL19xDUDg0zCSDNRuE1DzVNsIx0jtIA

**Comment by hallya** [comment]
Any news on this subject ? I'm facing the same issue

**Reply by devinrhode2** [reply]
@hallya did you try using latest RHF and nightly build of TS?

**Reply by Aaronius** [reply]
@devinrhode2 I did. I still get the same thing.

**Comment by Aaronius** [comment]
I was surprised to see that the return value from `formMethods.getValues()` was `UnpackNestedValue<TFieldNames>` rather than `TFieldNames`. Similarly, the callback I'm passing to `handleSubmit` receives `UnpackNestedValue<TFieldNames>` rather than `TFieldNames`.

**Comment by kunalbhagawati** [comment]
+1 to this. 
I had to work around by casting the object `defaultValues as DefaultValues<TFieldValues>`.

I am not able to make a non-hacky generically typed form yet because of this and `errors` not handling generics.
https://github.com/react-hook-form/react-hook-form/issues/7040

**Reply by vadimyen** [reply]
I have the same problem
